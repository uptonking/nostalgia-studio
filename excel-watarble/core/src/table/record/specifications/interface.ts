import {
  type CompositeSpecification,
  type ISpecVisitor,
} from '@datalking/pivot-entity';
import { type ISpecification } from '@datalking/pivot-entity';

import { type Record } from '../record';
import {
  type HasExtension,
  type HasFileType,
  type IsAttachmentEmpty,
} from './attachment.specification';
import { type BoolIsFalse, type BoolIsTrue } from './bool.specification';
import { type DateRangeEqual } from './date-range.specification';
import {
  type DateEqual,
  type DateGreaterThan,
  type DateGreaterThanOrEqual,
  type DateIsToday,
  type DateLessThan,
  type DateLessThanOrEqual,
} from './date.specification';
import {
  type NumberEqual,
  type NumberGreaterThan,
  type NumberGreaterThanOrEqual,
  type NumberLessThan,
  type NumberLessThanOrEqual,
} from './number.specification';
import { type ParentAvailableSpec } from './parent.specification';
import { type WithRecordAutoIncrement } from './record-auto-increment.specification';
import { type WithRecordCreatedAt } from './record-created-at.specification';
import { type WithRecordCreatedBy } from './record-created-by.specification';
import {
  type WithRecordId,
  type WithRecordIds,
} from './record-id.specification';
import { type WithRecordTableId } from './record-table-id.specification';
import { type WithRecordUpdatedAt } from './record-updated-at.specification';
import { type WithRecordUpdatedBy } from './record-updated-by.specification';
import { type WithRecordValues } from './record-values.specification';
import { type ReferenceEqual } from './reference.specification';
import { type SelectEqual, type SelectIn } from './select.specification';
import {
  type StringContain,
  type StringEndsWith,
  type StringEqual,
  type StringRegex,
  type StringStartsWith,
} from './string.specification';
import { type IsTreeRoot, type TreeAvailableSpec } from './tree.specification';

interface IRecordSpecVisitor {
  idEqual(s: WithRecordId): void;
  idsIn(s: WithRecordIds): void;
  tableIdEqual(s: WithRecordTableId): void;

  createdAt(s: WithRecordCreatedAt): void;
  createdBy(s: WithRecordCreatedBy): void;
  updatedAt(s: WithRecordUpdatedAt): void;
  updatedBy(s: WithRecordUpdatedBy): void;

  autoIncrement(s: WithRecordAutoIncrement): void;

  values(s: WithRecordValues): void;
}

interface IRecordValueVisitor {
  stringEqual(s: StringEqual): void;
  stringContain(s: StringContain): void;
  stringStartsWith(s: StringStartsWith): void;
  stringEndsWith(s: StringEndsWith): void;
  stringRegex(s: StringRegex): void;

  numberEqual(s: NumberEqual): void;
  numberGreaterThan(s: NumberGreaterThan): void;
  numberLessThan(s: NumberLessThan): void;
  numberGreaterThanOrEqual(s: NumberGreaterThanOrEqual): void;
  numberLessThanOrEqual(s: NumberLessThanOrEqual): void;

  dateEqual(s: DateEqual): void;
  dateGreaterThan(s: DateGreaterThan): void;
  dateLessThan(s: DateLessThan): void;
  dateGreaterThanOrEqual(s: DateGreaterThanOrEqual): void;
  dateLessThanOrEqual(s: DateLessThanOrEqual): void;
  dateIsToday(s: DateIsToday): void;

  dateRangeEqual(s: DateRangeEqual): void;

  selectEqual(s: SelectEqual): void;
  selectIn(s: SelectIn): void;

  boolIsTrue(s: BoolIsTrue): void;
  boolIsFalse(s: BoolIsFalse): void;

  referenceEqual(s: ReferenceEqual): void;

  treeAvailable(s: TreeAvailableSpec): void;
  isTreeRoot(s: IsTreeRoot): void;

  parentAvailable(s: ParentAvailableSpec): void;

  hasFileType(s: HasFileType): void;
  hasExtension(s: HasExtension): void;
  isAttachmentEmpty(s: IsAttachmentEmpty): void;
}

export type RecordCompositeSpecification = CompositeSpecification<
  Record,
  IRecordVisitor
>;

export type IRecordSpec = ISpecification<Record, IRecordVisitor>;

export type IRecordVisitor = IRecordSpecVisitor &
  IRecordValueVisitor &
  ISpecVisitor;
