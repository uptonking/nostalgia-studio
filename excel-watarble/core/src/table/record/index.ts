export * from './fixtures/index';
export * from './record-tree-query-model';
export * from './record.factory';
export * from './record';
export * from './record.query-model';
export * from './record.repository';
export * from './record.schema';
export * from './record.type';
export * from './specifications/index';
export * from './value-objects/index';
