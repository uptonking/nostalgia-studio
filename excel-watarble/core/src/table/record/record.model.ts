import { z } from 'zod';
import { autoIncrementQueryValue } from '../field/auto-increment-field.type';
import { fieldQueryValue } from '../field/index';
import { fieldIdSchema } from '../field/value-objects/field-id.schema';
import { recordIdSchema } from './value-objects/record-id.schema';

export const recordModel = z.object({
  id: recordIdSchema,
  createdAt: z.date(),
  updatedAt: z.date(),
  autoIncrement: autoIncrementQueryValue.optional(),
  values: z.record(fieldIdSchema, fieldQueryValue),
});

export type IRecordModel = z.infer<typeof recordModel>;
export type IRecordModelValues = IRecordModel['values'];
