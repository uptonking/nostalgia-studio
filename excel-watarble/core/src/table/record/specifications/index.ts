export * from './attachment.specification';
export * from './bool.specification';
export * from './date-range.specification';
export * from './date.specification';
export type { IRecordSpec, IRecordVisitor } from './interface';
export * from './number.specification';
export * from './parent.specification';
export * from './record-auto-increment.specification';
export * from './record-created-at.specification';
export * from './record-created-by.specification';
export * from './record-display-values.specification';
export * from './record-id.specification';
export * from './record-table-id.specification';
export * from './record-updated-at.specification';
export * from './record-updated-by.specification';
export * from './record-values.specification';
export * from './reference.specification';
export * from './select.specification';
export * from './string.specification';
export * from './tree.specification';
