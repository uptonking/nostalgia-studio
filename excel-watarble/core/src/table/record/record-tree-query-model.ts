import { type TreeField } from '../field/index';
import { type IQueryTreeRecords } from './record.type';
import { type IRecordSpec } from './specifications/index';

export interface IRecordTreeQueryModel {
  findTrees(
    tableId: string,
    field: TreeField,
    spec: IRecordSpec,
  ): Promise<IQueryTreeRecords>;
}
