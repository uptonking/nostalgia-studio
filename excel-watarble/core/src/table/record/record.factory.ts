import { type Result } from 'oxide.ts';

import { and } from '@datalking/pivot-entity';

import { type TableSchemaIdMap } from '../value-objects/index';
import { Record } from './record';
import { type IQueryRecordSchema, type Records } from './record.type';
import {
  WithDisplayValues,
  WithRecordCreatedAt,
  WithRecordCreatedBy,
  WithRecordCreatedByProfile,
  WithRecordId,
  WithRecordTableId,
  WithRecordUpdatedAt,
  WithRecordValues,
} from './specifications/index';
import { type RecordCompositeSpecification } from './specifications/interface';
import { WithRecordAutoIncrement } from './specifications/record-auto-increment.specification';
import {
  WithRecordUpdatedBy,
  WithRecordUpdatedByProfile,
} from './specifications/record-updated-by.specification';

export class RecordFactory {
  static create(
    ...specs: RecordCompositeSpecification[]
  ): Result<Record, string>;
  static create(spec: RecordCompositeSpecification): Result<Record, string>;

  static create(
    spec: RecordCompositeSpecification | RecordCompositeSpecification[],
  ): Result<Record, string> {
    if (Array.isArray(spec)) {
      return and(...spec)
        .unwrap()
        .mutate(Record.empty());
    }
    return spec.mutate(Record.empty());
  }

  static fromQueryRecords(
    rs: IQueryRecordSchema[],
    schema: TableSchemaIdMap,
  ): Records {
    return rs.map((r) => this.fromQuery(r, schema).unwrap());
  }

  static fromQuery(
    r: IQueryRecordSchema,
    schema: TableSchemaIdMap,
  ): Result<Record, string> {
    let spec = WithRecordId.fromString(r.id)
      .and(WithRecordTableId.fromString(r.tableId).unwrap())
      .and(WithRecordCreatedAt.fromString(r.createdAt))
      .and(WithRecordCreatedBy.fromString(r.createdBy))
      .and(new WithRecordCreatedByProfile(r.createdByProfile))
      .and(WithRecordUpdatedAt.fromString(r.updatedAt))
      .and(WithRecordUpdatedBy.fromString(r.updatedBy))
      .and(new WithRecordUpdatedByProfile(r.updatedByProfile))
      .and(WithRecordValues.fromObject(schema, r.values))
      .and(WithDisplayValues.from(r.displayValues));

    if (typeof r.autoIncrement === 'number') {
      spec = spec.and(new WithRecordAutoIncrement(r.autoIncrement));
    }

    return this.create(spec);
  }
}
