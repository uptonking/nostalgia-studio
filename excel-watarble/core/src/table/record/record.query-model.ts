import { type Option } from 'oxide.ts';
import { type ViewId } from '../view/index';
import { type IQueryRecords, type IQueryRecordSchema } from './record.type';
import { type IRecordSpec } from './specifications/index';

export interface IRecordQueryModel {
  findOne(
    tableId: string,
    spec: IRecordSpec,
  ): Promise<Option<IQueryRecordSchema>>;
  findOneById(tableId: string, id: string): Promise<Option<IQueryRecordSchema>>;
  find(
    tableId: string,
    viewId: ViewId | undefined,
    spec: IRecordSpec | null,
  ): Promise<IQueryRecords>;
  findAndCount(
    tableId: string,
    viewId: ViewId | undefined,
    spec: IRecordSpec | null,
  ): Promise<{ records: IQueryRecords; total: number }>;
}
