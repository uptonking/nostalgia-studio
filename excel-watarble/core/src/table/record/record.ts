import { DateVO } from '@datalking/pivot-entity';

import { type ICollaboratorProfile } from '../field/collaborator-field.type';
import {
  type TableId,
  type TableSchema,
  type TableSchemaIdMap,
} from '../value-objects/index';
import { RecordFactory } from './record.factory';
import {
  type IInternalRecordValues,
  type IMutateRecordValueSchema,
  type RecordAllValues,
} from './record.schema';
import { createRecordInputs } from './record.utils';
import {
  WithRecordId,
  WithRecordTableId,
  WithRecordValues,
} from './specifications/index';
import { type RecordCompositeSpecification } from './specifications/interface';
import { RecordId, RecordValues } from './value-objects/index';
import { RecordDisplayValues } from './value-objects/record-display-values.vo';

export class Record {
  public id: RecordId = RecordId.create();
  public tableId!: TableId;
  public values: RecordValues = RecordValues.empty();
  public displayValues?: RecordDisplayValues = RecordDisplayValues.empty();
  public createdAt: DateVO = DateVO.now();
  public createdBy!: string;
  public createdByProfile: ICollaboratorProfile | null = null;
  public updatedBy!: string;
  public updatedByProfile: ICollaboratorProfile | null = null;
  public updatedAt: DateVO = DateVO.now();
  public autoIncrement?: number;

  // eslint-disable-next-line @typescript-eslint/no-empty-function
  private constructor() {}

  static empty() {
    const record = new Record();

    return record;
  }

  get internalValuesJSON(): IInternalRecordValues {
    return {
      id: this.id.value,
      created_at: this.createdAt.value.toISOString(),
      created_by: this.createdBy,
      created_by_profile: this.createdByProfile,
      updated_at: this.updatedAt.value.toISOString(),
      updated_by: this.updatedBy,
      updated_by_profile: this.updatedByProfile,
      auto_increment: this.autoIncrement,
      display_values: this.displayValues?.values,
    };
  }

  get valuesJSON(): RecordAllValues {
    return Object.assign({}, this.internalValuesJSON, this.values.valueJSON);
  }

  updateRecord(
    schema: TableSchema,
    value: IMutateRecordValueSchema,
  ): RecordCompositeSpecification {
    const inputs = createRecordInputs(schema, value);
    const spec = WithRecordValues.fromArray(inputs);

    spec.mutate(this);

    return spec;
  }

  duplicate(schema: TableSchemaIdMap): Record {
    return RecordFactory.create(
      new WithRecordId(RecordId.create())
        .and(new WithRecordTableId(this.tableId))
        .and(new WithRecordValues(this.values.duplicate(schema))),
    ).unwrap();
  }
}
