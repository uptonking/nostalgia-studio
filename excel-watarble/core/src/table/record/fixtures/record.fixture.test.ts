import { NumberFieldValue, StringFieldValue } from '../../field/index';
import { NumberEqual, StringEqual } from '../specifications/index';
import { type RecordCompositeSpecification } from '../specifications/interface';
import { createTestRecord } from './record.fixture';

beforeAll(() => {
  vi.setSystemTime(new Date(2022, 1, 1));
});

test.each<RecordCompositeSpecification[][]>([
  [[new StringEqual('hello', new StringFieldValue('world'))]],
  [[new NumberEqual('hello', new NumberFieldValue(1))]],
  [
    [
      new StringEqual('hello', new StringFieldValue('world')),
      new NumberEqual('number', new NumberFieldValue(1)),
    ],
  ],
])('createTestRecord', (specs) => {
  const record = createTestRecord(...specs);
  expect(record).toMatchSnapshot();
});
