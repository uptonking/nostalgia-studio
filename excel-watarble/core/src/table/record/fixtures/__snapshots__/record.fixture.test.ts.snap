// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createTestRecord 1`] = `
Record {
  "autoIncrement": undefined,
  "createdAt": c {
    "props": {
      "value": 2022-02-01T00:00:00.000Z,
    },
  },
  "createdBy": undefined,
  "createdByProfile": null,
  "displayValues": RecordDisplayValues {
    "props": {},
  },
  "id": RecordId {
    "props": {
      "value": "record",
    },
  },
  "tableId": TableId {
    "props": {
      "value": "tableId",
    },
  },
  "updatedAt": c {
    "props": {
      "value": 2022-02-01T00:00:00.000Z,
    },
  },
  "updatedBy": undefined,
  "updatedByProfile": null,
  "values": RecordValues {
    "props": Map {
      "hello" => StringFieldValue {
        "props": {
          "value": "world",
        },
      },
    },
  },
}
`;

exports[`createTestRecord 2`] = `
Record {
  "autoIncrement": undefined,
  "createdAt": c {
    "props": {
      "value": 2022-02-01T00:00:00.000Z,
    },
  },
  "createdBy": undefined,
  "createdByProfile": null,
  "displayValues": RecordDisplayValues {
    "props": {},
  },
  "id": RecordId {
    "props": {
      "value": "record",
    },
  },
  "tableId": TableId {
    "props": {
      "value": "tableId",
    },
  },
  "updatedAt": c {
    "props": {
      "value": 2022-02-01T00:00:00.000Z,
    },
  },
  "updatedBy": undefined,
  "updatedByProfile": null,
  "values": RecordValues {
    "props": Map {
      "hello" => NumberFieldValue {
        "props": {
          "value": 1,
        },
      },
    },
  },
}
`;

exports[`createTestRecord 3`] = `
Record {
  "autoIncrement": undefined,
  "createdAt": c {
    "props": {
      "value": 2022-02-01T00:00:00.000Z,
    },
  },
  "createdBy": undefined,
  "createdByProfile": null,
  "displayValues": RecordDisplayValues {
    "props": {},
  },
  "id": RecordId {
    "props": {
      "value": "record",
    },
  },
  "tableId": TableId {
    "props": {
      "value": "tableId",
    },
  },
  "updatedAt": c {
    "props": {
      "value": 2022-02-01T00:00:00.000Z,
    },
  },
  "updatedBy": undefined,
  "updatedByProfile": null,
  "values": RecordValues {
    "props": Map {
      "hello" => StringFieldValue {
        "props": {
          "value": "world",
        },
      },
      "number" => NumberFieldValue {
        "props": {
          "value": 1,
        },
      },
    },
  },
}
`;
